File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/CPP/ValidateBinarySearchTree.cpp
// In order traversal 

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool CheckSorted(vector<int> &vnumber){
        bool sorted=true;
        if(vnumber.size() <=1)
            return sorted;
        
        int prev=vnumber[0];
        for( int i=1; i<vnumber.size(); i++){
            if(prev>=vnumber[i])
                return !sorted;
            prev=vnumber[i];
        }
        return sorted;
    }
    
    void InOrderTraverse( TreeNode* node, vector<int> & vnum){
        if( node->left !=NULL)
            InOrderTraverse(node->left, vnum);
        // visit node
        vnum.push_back(node->val);
        
        if( node->right!=NULL)
            InOrderTraverse(node->right,vnum);
    }
    
    bool isValidBST(TreeNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if( root==NULL)
            return true;
        vector<int> result_num;
        InOrderTraverse(root,result_num);
        bool validBST=CheckSorted(result_num);
        return validBST;
        
    }

};

File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Python/ValidateBinarySearchTree.py
#Given a binary tree, determine if it is a valid binary search tree (BST).
#
#Assume a BST is defined as follows:
#
#The left subtree of a node contains only nodes with keys less than the node's key.
#The right subtree of a node contains only nodes with keys greater than the node's key.
#Both the left and right subtrees must also be binary search trees.


# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
    def isValidBST(self, root):
        tree_nodes = [] 
        # inorder traversal see is the output list sorted
        self.inorderTraversal(root, tree_nodes)

        # python all(iterable) built-in function
        return all( tree_nodes[idx] < tree_nodes[idx+1] for idx in range(len(tree_nodes) -1) ) 

    def inorderTraversal(self, node, tree_nodes):
        if node == None:    return
        
        self.inorderTraversal( node.left, tree_nodes)
        tree_nodes.append( node.val)
        self.inorderTraversal( node.right, tree_nodes)


                    
